"
This R script performs differential gene expression analysis using Bioconductor DESeq2 package.
Read: https://bioconductor.org/packages/release/bioc/html/DESeq2.html
Read the manual for additional customizations.
The script assumes that the DESeq2 and ggplot2 libraries are already installed.

Input data file must be an output generated by featureCounts in .csv format and contain Geneid column and read counts in subsequent columns.
Make sure the column names of datasets are simplified by removing the path.
In addtion to the input .csv file, exclude_genes.csv must be in the working directory. 
This file contains gene names of rRNAs, tRNAs, noncoding RNAs and pseudogenes to be substracted from the input .csv file.
Generate this in Biomart and download. Column name of the gene names column must be 'Geneid'.
"

#### Loading required libraries -------------
library(DESeq2)
library(ggplot2)

#### Setting up the working environment -------------

## Set-up the working directory.
working_dir <- "/path/to/working/directory/"      ## This path should contain the input file. Ex: gene.csv
setwd(working_dir)

#### Data import and preprocessing -------------

## Import data from FeatureCounts output files.
countdata <- read.csv("gene.csv", header=TRUE, sep = ",")
colnames(countdata)

## To remove rRNA/tRNA/noncoding/pseudogenes
exclude_genes <- read.csv("exclude_genes.csv", header=TRUE, sep = ",")
colnames(exclude_genes)
countdata <- anti_join(countdata, exclude_genes, by = "Geneid")
head(countdata)

## Assign Geneid as the row name column
rownames(countdata) <- countdata[,1]
countdata[,1] <- NULL
head(countdata)

## To rearrange columns (if needed)
countdata <- countdata[,c("column_1", "column_2", "column_3", "column_7", "column_8", "column_9")]
colnames(countdata)

## Convert pre-processed data into a matrix.
countdata <- as.matrix(countdata)
colnames(countdata)

## Assign condition.
(genotype <- factor(c(rep("genotype_1", 3), rep("genotype_2", 3))))

#### DESeq2 analysis -------------

## Create a coldata dataframe.
(coldata <- data.frame(row.names=colnames(countdata), genotype))

## Instantiate the DESeqDataSet.
dds <- DESeqDataSetFromMatrix(countData=countdata, colData=coldata, design=~genotype)
dds

## Run the DESeq2 pipeline.
dds <- DESeq(dds)

## Get differential expression results.
resultsNames(dds)
res <- results(dds, contrast = c("genotype", "genotype_1", "genotype_2"))
head(res)

## Order by adjusted p-value.
res <- res[order(res$padj), ]
head(res)

## Add gene names and write results into a .csv file.
resdata <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdata)[1] <- "Gene"
head(resdata)
write.csv(resdata, file = "DESeq2_genotype1_vs_genotype2_Results.csv")

#### PCA/Principle Component Analysis -------------

## To transform count data using the concept of variance stabilizing transformation (VST).
vsd <- varianceStabilizingTransformation(dds)
head(vsd)

## Use plotPCA function to generate the PCA plot and generate first two PCs, PC1 and PC2.
pcaData <- plotPCA(vsd, intgroup=c("genotype"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
percentVar

## Customize the plot using ggplot2.
pcaplot <- ggplot(pcaData, aes(x = PC1, y = PC2, color=genotype)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  theme(axis.title.x = element_text(face = "bold", size = 12),
        axis.text.x = element_text(size = 10)) +
  theme(axis.title.y = element_text(face = "bold", size = 12),
        axis.text.y = element_text(size = 10)) +
  theme(legend.position="bottom")

pcaplot

pdf("PCA_plot.pdf", width=5, height = 5)
pcaplot
dev.off()